<a href="https://cognitiveclass.ai"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Models with Keras</font></h1>

## Introduction

As we discussed in the videos, despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library. 

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.

## Table of Contents

<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 3>
    
1. <a href="#item31">Download and Clean Dataset</a>  
2. <a href="#item32">Import Keras</a>  
3. <a href="#item33">Build a Neural Network</a>  
4. <a href="#item34">Train and Test the Network</a>  

</font>
</div>

<a id="item31"></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [3]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's download the data and read it into a <em>pandas</em> dataframe.

In [4]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head(100)

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0    540.0                 0.0      0.0  162.0               2.5   
1    540.0                 0.0      0.0  162.0               2.5   
2    332.5               142.5      0.0  228.0               0.0   
3    332.5               142.5      0.0  228.0               0.0   
4    198.6               132.4      0.0  192.0               0.0   
..     ...                 ...      ...    ...               ...   
95   425.0               106.3      0.0  153.5              16.5   
96   425.0               106.3      0.0  151.4              18.6   
97   375.0                93.8      0.0  126.6              23.4   
98   475.0               118.8      0.0  181.1               8.9   
99   469.0               117.2      0.0  137.8              32.2   

    Coarse Aggregate  Fine Aggregate  Age  Strength  
0             1040.0           676.0   28     79.99  
1             1055.0           676.0   28     61.89  
2              932.0           594.0  270     40.27  
3              932.0           594.0  365     41.05  
4              978.4           825.5  360     44.30  
..               ...             ...  ...       ...  
95             852.1           887.1    7     49.20  
96             936.0           803.7    7     46.80  
97             852.1           992.6    7     45.70  
98             852.1           781.5    7     55.60  
99             852.1           840.5    7     54.90  

[100 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [5]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [6]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [9]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [11]:
concrete_data_columns = concrete_data.columns
print(concrete_data_columns)

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Index(['Cement', 'Blast Furnace Slag', 'Fly Ash', 'Water', 'Superplasticizer',
       'Coarse Aggregate', 'Fine Aggregate', 'Age', 'Strength'],
      dtype='object')


<a id="item2"></a>

Let's do a quick sanity check of the predictors and the target dataframes.

In [8]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [9]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [17]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.

In [19]:
n_cols = predictors_norm.shape[1] # number of predictors

<a id="item1"></a>

<a id='item32'></a>

# Import Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

#### Let's go ahead and import the Keras library

In [20]:
import keras

Using TensorFlow backend.
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/home/jupyterlab/conda/envs/python/lib/python3.6/site-packages/tensorflow/python/framework/dtypes.py:521: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/home/jupyterlab/con

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [21]:
from keras.models import Sequential
from keras.layers import Dense

<a id='item33'></a>

## Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [31]:
# define regression model
def regression_model(model_optimizer, model_loss_function):
    # create model
    model = Sequential()
    model.add(Dense(50, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer=model_optimizer, loss=model_loss_function)
    return model

The above function create a model that has two hidden layers, each of 50 hidden units.

<a id="item4"></a>

<a id='item34'></a>

## Train and Test the Network

Let's call the function now to create our model.

In [37]:
# build the model
# model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.

In [55]:
# rebuild the model to initialize
# (otherwise the network is already trained)
model = regression_model('adam','mse')

# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 1s - loss: 1621.5931 - val_loss: 1132.8650
Epoch 2/100
 - 0s - loss: 1473.0159 - val_loss: 995.6872
Epoch 3/100
 - 0s - loss: 1226.7736 - val_loss: 785.7036
Epoch 4/100
 - 0s - loss: 885.4861 - val_loss: 529.2413
Epoch 5/100
 - 0s - loss: 534.0571 - val_loss: 319.9639
Epoch 6/100
 - 0s - loss: 316.0563 - val_loss: 213.5261
Epoch 7/100
 - 0s - loss: 243.8519 - val_loss: 177.1206
Epoch 8/100
 - 0s - loss: 222.3815 - val_loss: 162.5105
Epoch 9/100
 - 0s - loss: 207.8610 - val_loss: 157.1127
Epoch 10/100
 - 0s - loss: 196.0746 - val_loss: 153.3979
Epoch 11/100
 - 0s - loss: 187.7258 - val_loss: 147.5696
Epoch 12/100
 - 1s - loss: 180.1856 - val_loss: 145.1869
Epoch 13/100
 - 0s - loss: 174.2892 - val_loss: 143.6209
Epoch 14/100
 - 0s - loss: 169.4501 - val_loss: 141.7609
Epoch 15/100
 - 0s - loss: 165.4788 - val_loss: 141.0862
Epoch 16/100
 - 0s - loss: 161.0751 - val_loss: 138.9705
Epoch 17/100
 - 0s - loss: 157.3750 - val_loss:

In [54]:
# rebuild the model to initialize
# (otherwise the network is already trained)
model = regression_model('adam','mean_squared_error')

# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 1s - loss: 1636.2179 - val_loss: 1128.3022
Epoch 2/100
 - 0s - loss: 1483.7692 - val_loss: 984.3821
Epoch 3/100
 - 0s - loss: 1240.5062 - val_loss: 766.8934
Epoch 4/100
 - 0s - loss: 902.5273 - val_loss: 502.0000
Epoch 5/100
 - 0s - loss: 545.0988 - val_loss: 281.9602
Epoch 6/100
 - 0s - loss: 321.1394 - val_loss: 183.0737
Epoch 7/100
 - 0s - loss: 244.9031 - val_loss: 171.9075
Epoch 8/100
 - 0s - loss: 222.0271 - val_loss: 173.5725
Epoch 9/100
 - 0s - loss: 209.5724 - val_loss: 171.4300
Epoch 10/100
 - 0s - loss: 199.7520 - val_loss: 171.4564
Epoch 11/100
 - 0s - loss: 192.9677 - val_loss: 168.1076
Epoch 12/100
 - 0s - loss: 186.4067 - val_loss: 166.7071
Epoch 13/100
 - 0s - loss: 180.8813 - val_loss: 163.3912
Epoch 14/100
 - 0s - loss: 176.2845 - val_loss: 162.2557
Epoch 15/100
 - 0s - loss: 171.6979 - val_loss: 162.0833
Epoch 16/100
 - 0s - loss: 168.0071 - val_loss: 162.4971
Epoch 17/100
 - 0s - loss: 165.1322 - val_loss:

<strong>You can refer to this [link](https://keras.io/models/sequential/) to learn about other functions that you can use for prediction or evaluation.</strong>

In [53]:
# rebuild the model to initialize
# (otherwise the network is already trained)
model = regression_model('sgd','mean_squared_error')

# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 1s - loss: nan - val_loss: nan
Epoch 2/100
 - 0s - loss: nan - val_loss: nan
Epoch 3/100
 - 0s - loss: nan - val_loss: nan
Epoch 4/100
 - 0s - loss: nan - val_loss: nan
Epoch 5/100
 - 0s - loss: nan - val_loss: nan
Epoch 6/100
 - 0s - loss: nan - val_loss: nan
Epoch 7/100
 - 0s - loss: nan - val_loss: nan
Epoch 8/100
 - 0s - loss: nan - val_loss: nan
Epoch 9/100
 - 0s - loss: nan - val_loss: nan
Epoch 10/100
 - 0s - loss: nan - val_loss: nan
Epoch 11/100
 - 0s - loss: nan - val_loss: nan
Epoch 12/100
 - 0s - loss: nan - val_loss: nan
Epoch 13/100
 - 0s - loss: nan - val_loss: nan
Epoch 14/100
 - 0s - loss: nan - val_loss: nan
Epoch 15/100
 - 0s - loss: nan - val_loss: nan
Epoch 16/100
 - 0s - loss: nan - val_loss: nan
Epoch 17/100
 - 0s - loss: nan - val_loss: nan
Epoch 18/100
 - 0s - loss: nan - val_loss: nan
Epoch 19/100
 - 0s - loss: nan - val_loss: nan
Epoch 20/100
 - 0s - loss: nan - val_loss: nan
Epoch 21/100
 - 0s - lo

Feel free to vary the following and note what impact each change has on the model's performance:

1. Increase or decreate number of neurons in hidden layers
2. Add more hidden layers
3. Increase number of epochs

In [52]:
# rebuild the model to initialize
# (otherwise the network is already trained)
model = regression_model('sgd','KLD')

# fit the model
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Train on 721 samples, validate on 309 samples
Epoch 1/100
 - 1s - loss: 0.2252 - val_loss: 0.0206
Epoch 2/100
 - 0s - loss: 0.0105 - val_loss: 0.0091
Epoch 3/100
 - 0s - loss: 0.0033 - val_loss: 0.0054
Epoch 4/100
 - 0s - loss: 0.0016 - val_loss: 0.0043
Epoch 5/100
 - 0s - loss: 0.0011 - val_loss: 0.0036
Epoch 6/100
 - 0s - loss: 8.3099e-04 - val_loss: 0.0030
Epoch 7/100
 - 0s - loss: 6.1623e-04 - val_loss: 0.0026
Epoch 8/100
 - 0s - loss: 4.4432e-04 - val_loss: 0.0022
Epoch 9/100
 - 0s - loss: 2.9226e-04 - val_loss: 0.0019
Epoch 10/100
 - 0s - loss: 2.1128e-04 - val_loss: 0.0018
Epoch 11/100
 - 0s - loss: 1.6262e-04 - val_loss: 0.0016
Epoch 12/100
 - 0s - loss: 1.0764e-04 - val_loss: 0.0015
Epoch 13/100
 - 0s - loss: 6.9518e-05 - val_loss: 0.0014
Epoch 14/100
 - 0s - loss: 5.5298e-05 - val_loss: 0.0014
Epoch 15/100
 - 0s - loss: 4.1193e-05 - val_loss: 0.0013
Epoch 16/100
 - 0s - loss: 2.7204e-05 - val_loss: 0.0012
Epoch 17/100
 - 0s - loss: 1.3327e-05 - val_loss: 0.0012
Epoch 18/100
 

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!

This notebook is part of a course on **Coursera** called *Introduction to Deep Learning & Neural Networks with Keras*. If you accessed this notebook outside the course, you can take this course online by clicking [here](https://cocl.us/DL0101EN_Coursera_Week3_LAB1).

<hr>

Copyright &copy; 2019 [IBM Developer Skills Network](https://cognitiveclass.ai/?utm_source=bducopyrightlink&utm_medium=dswb&utm_campaign=bdu). This notebook and its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/).